Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -64,151 +60,151 @@ class SerializerTest : public ::testing::Test {
   }
   Serializer serializer;
 
-  void ExpectRoundTrip(const FieldValue& model,
-                       const std::vector<uint8_t>& bytes,
-                       FieldValue::Type type) {
+  void ExpectRoundTrip(const FieldValue& model, FieldValue::Type type) {
+    google::firestore::v1beta1::Value proto = ConvertModelToProto(model);
+
+    // First, serialize model with our (nanopb based) serializer, then
+    // deserialize the resulting bytes with libprotobuf and ensure the result is
+    // the same as the expected proto.
+    ExpectSerializationRoundTrip(model, proto, type);
+
+    // Next, serialize proto with libprotobuf, then deserialize the resulting
+    // bytes with our (nanopb based) deserializer and ensure the result is the
+    // same as the expected model.
+    ExpectDeserializationRoundTrip(model, proto, type);
+  }
+
+ private:
+  google::firestore::v1beta1::Value ConvertModelToProto(
```

Comment:
Generating the expected values from the source models will end up largely duplicating the serializer's logic and makes these tests too magical.

On the other platforms (and in the old Objective-C versions of these), the expected values were [assembled by hand](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Example/Tests/Remote/FSTSerializerBetaTests.mm#L123). In the interest of keeping the tests simpler I think that's called for here too.

Reply:
Manually assembling them completely by hand doesn't seem like a win to me. (My first version of this did that, but I found I was duplicating stuff over and over, so I refactored it out into this function.) This function doesn't really do anything that the existing objc tests didn't; they just move it to a common location. (It does recurse, but I think that's about the only additional thing it does by itself.)

> ... duplicating the serializer's logic ...

Yeah, there's some truth to that. Though I'd argue that the existing objc tests largely do this too. (Obj-c builds up a proto based on the same values that are in the model, and then asks the serializer to build a proto based on the model and compares them. In both cases, libprotobuf is used to do so. At least with C++ there's a libprotobuf vs nanopb wrinkle.)

Still, I've inlined the map test (since it's the most interesting) into 3 different options.

Option 1: as is (ie. with this function.)
Option 2: fully inlined. (Which I think is pretty ugly. YMMV.)
Option 3: Somewhere in between. Requires additional functions like EncodeString() be added. (Or exposed as public from the implementation.) I think this most closely resembles the objc tests.

My preference (in order) is 1, 3, 2, but am willing to go with any of them.